From: SeongJae Park <sj@kernel.org>
Date: Sun, 6 Jul 2025 10:43:48 -0700
Subject: [PATCH] mm/damon/core: fix a race

Once the control is completed, damon_call() caller can free the control
object, and hence following list_del(&control->list) shows correupted
list.  Do the completion after list_del() is done.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 38c37938ba7f..01e62bdcdd67 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2404,10 +2404,10 @@ static void kdamond_call(struct damon_ctx *ctx, bool cancel)
 			ret = control->fn(control->data);
 			control->return_code = ret;
 		}
-		complete(&control->completion);
 		mutex_lock(&ctx->call_controls_lock);
 		list_del(&control->list);
 		mutex_unlock(&ctx->call_controls_lock);
+		complete(&control->completion);
 	}
 }
 
-- 
2.39.5

